File: src/main/java/io/github/richardson/gameplay/Background.java
package io.github.richardson.gameplay;

import processing.core.PApplet;
import processing.core.PGraphics;

public class Background {
    private final PApplet p;
    private final PGraphics bg;

    public Background(PApplet p, int c1, int c2) {
        this.p = p;
        this.bg = p.createGraphics(p.width, p.height);
        this.bg.beginDraw();
        this.bg.noFill();
        for (int i = 0; i <= p.height; i++) {
            this.bg.stroke(p.lerpColor(c1, c2, (float) Math.sqrt((float) i / p.height)));
            this.bg.line(0, i, p.width, i);
        }
        this.bg.endDraw();
    }

    public void render() {
        p.image(bg, 0, 0);
        //java.lang.System.out.println("rendering background (" + p.millis() + "ms)");
    }
}

File: src/gameplay/Background.kt
package gameplay

import p
import processing.core.PApplet
import processing.core.PGraphics
import kotlin.math.sqrt

class Background(c1: Int, c2: Int) {
    private val bg: PGraphics = p.createGraphics(p.displayWidth, p.displayHeight)

    init { // TODO: eliminate color banding
        this.bg.beginDraw()
        this.bg.noFill()
        for (i in 0..p.displayHeight) {
            this.bg.stroke(p.lerpColor(c1, c2, sqrt((i.toFloat() / p.displayHeight).toDouble()).toFloat()))
            this.bg.line(0f, i.toFloat(), p.displayWidth.toFloat(), i.toFloat())
        }
        this.bg.endDraw()
    }

    fun render() {
        p.imageMode(PApplet.CORNER)
        p.image(bg, 0f, 0f, p.width.toFloat(), p.height.toFloat())
    }
}